pkg/virt-controller/watch/migration.go

```diff
@@ -26,111 +26,122 @@ func NewMigrationControllerWithListWatch(migrationService services.VMService, _
 	return kubecli.NewController(lw, queue, &v1.Migration{}, NewMigrationControllerFunc(migrationService))
 }
 
-func NewMigrationControllerFunc(migrationService services.VMService) kubecli.ControllerFunc {
-	return func(store cache.Store, queue workqueue.RateLimitingInterface, key interface{}) {
-		// Fetch the latest Migration state from cache
-		obj, exists, err := store.GetByKey(key.(string))
-
-		if err != nil {
-			queue.AddRateLimited(key)
-			return
-		}
-		if exists {
-			var migration *v1.Migration = obj.(*v1.Migration)
-			logger := logging.DefaultLogger().Object(migration)
-			if migration.Status.Phase == v1.MigrationUnknown {
-				// Copy migration for future modifications
-				migrationCopy, err := copy(migration)
+func NewMigrationControllerFunc(vmService services.VMService) kubecli.ControllerDispatch {
+
+	dispatch := MigrationDispatch{
+		vmService: vmService,
+	}
+	return &dispatch
+}
+
+type MigrationDispatch struct {
+	vmService services.VMService
+}
+
+func (md *MigrationDispatch) Execute(store cache.Store, queue workqueue.RateLimitingInterface, key interface{}) {
+
+	// Fetch the latest Migration state from cache
+	obj, exists, err := store.GetByKey(key.(string))
+
+	if err != nil {
+		queue.AddRateLimited(key)
+		return
+	}
+	if exists {
+		var migration *v1.Migration = obj.(*v1.Migration)
+		logger := logging.DefaultLogger().Object(migration)
+		if migration.Status.Phase == v1.MigrationUnknown {
+			// Copy migration for future modifications
+			migrationCopy, err := copy(migration)
+			if err != nil {
+				logger.Error().Reason(err).Msg("could not copy migration object")
+				queue.AddRateLimited(key)
+				return
+			}
+			// Fetch vm which we want to migrate
+			vm, exists, err := md.vmService.FetchVM(migration.Spec.Selector.Name)
+			if err != nil {
+				logger.Error().Reason(err).Msgf("fetching the vm %s failed", migration.Spec.Selector.Name)
+				queue.AddRateLimited(key)
+				return
+			}
+			if !exists {
+				logger.Info().Msgf("VM with name %s does not exist, marking migration as failed", migration.Spec.Selector.Name)
+				migrationCopy.Status.Phase = v1.MigrationFailed
+				// TODO indicate why it was set to failed
+				err := md.vmService.UpdateMigration(migrationCopy)
 				if err != nil {
-					logger.Error().Reason(err).Msg("could not copy migration object")
+					logger.Error().Reason(err).Msg("updating migration state failed")
```

Comment:
The log message "updating migration state failed" is now used 8 times in this module. It would be better to use more specific messages (especially containing the migration/VM name if it's present) if possible.